README
Description
-----------

Rdopt is a program to generate customized DCT quantization
tables for use in JPEG compression. The input is an image
file (raw row-major bytes or pgm or ppm). The program runs an
optimization algorithm and then prompts the user to enter
compression or SNR/PSNR specifications whereupon it outputs
a Quantization Table that when used in JPEG compression with
customized Huffman Tables (eg., Independent JPEG group's
cjpeg program with the options cjpeg -optimize -qtables <Q-FILE>)
compresses the image to within those specifications optimally.

NOTE: If you avoid reading the CHANGES file, the main thing you
need to know is:
  use flag "-method lagrangian" for faster speed
  (except when.. heck, read the CHANGES file, or the
  "MORE INFO" part of this file)

Compiling
---------

Copy the appropriate makefile.{DEC,HP,RS6000,PENTIUM,SUN} file
onto makefile. Edit makefile: especially, set BINPATH to wherever
you want to put the executable. Do "make" and pray! 


Usage
-----


rdopt [flags] {-im imfile [image_flags] | -hist histfile}
where,
flags can be:

	 -v: verbose output: the more -v's, the more verbose the output
	 -silent: Normal stderr output will be suppressed
	 -cmdfile cfile: after optimization, read commands from cfile instead of stdin
	 -height H: nominal height of the image is H (default 512)
	 -width W: nominal width of the image is W (default 512)
	 -planes N: there are N color planes in the image (default 1)
	            these color planes are numbered 0..(N-1)
	 -numtables t: t tables will be used for quantization.
	               1 table each for color planes numbered 0..(t-2)
	               Color planes numbered (t-1) thru (N-1) will be
	               quantized together (default 1).
	 -clampDC d: don't consider qtable entries greater than d
	             for the DC coefficient (default 255 or QTABENTRYMAX for >8-bit prec)
	 -dontclampDC: consider all qtable entries for DC too
	 -dcdpcm:  use an approximation of dpcm coding done for
	            the DC coef to estimate its entropy.
	 -bppmax b: consider bits per pixel values upto b (default 1.0)
	 -bppscale B: discretize bpp by the integer B (default 5000)
	              the slowness of the program grows linearly with b*B
	 -thresh T: try thresholds upto T/2 greater than q/2. (default 0)
	 -weights n cwfile: for unit n, the errors in various DCT coefficients
	                  are to be weighted according to the weights
	                  listed in RMWSS order in cwfile. these weights
	                  will be normalized so that they add to 64
	                  Useful for using perceptually weighted distortion measures
	 -pweights w0[,w1,..]: for aggregate error,  unit n
	               will have weight wn. Missing weights will be set
	               to the last specified weight. weights will be
	               normalized to add up to numtables
	 -subsamp n h w: color plane number n is to be subsampled to height H/h and 
	                 width W/w. defaults are h=w=1 for each
	                 color plane.
	 -insubsamp n h w: color plane number n in input file has height H/h and 
	                 width W/w. defaults are h=w=1 for each
	                 color plane. H(W) must be divisible by h(w)
	 -mintable n fname: qtable # n has to have entries no lesser
	                    than those in the table listed in
	                    row-major white-space-separated (RMWSS)
	                    order in the file fname
	 -maxtable n fname: qtable # n has to have entries no more
	                    than those in the table listed in
	                    RMWSS order in the file fname
	 -correct c: find actual bpp at predicted-bpp-c, apply correction thereafter
	 -plot pfile: dump a plot of bpp-psnr pairs in the
	              file pfile
	 -points n: use n points in -plot (default 20)
	 -pbppmax b: plot bpp range 0-b (default min of bppmax,1.5)
	 -bppplane n: useful for images with > 1 planes. Use plane
	             number n to calculate # of pixels for reporting
	             bits per pixel. Default is to use the sum of
	           #pixels in each plane
	 -errfile fname: send stderr to file fname (can be - for stdout)

	 ** Obscure flags that you shouldn't really need:

	 -stats: histograms will be dumped in the file HISTOGRAM
	 -onlystats: only histogram-dumping, no optimization
	 -mapq: qentries tried will increase logarithmically
	        This is useful for speeding up rdopt with 12+ bit samples

** Either the -im flag or the -hist flag must be present. In case
of -hist, the file histfile must contain a histogram generated
by rdopt -stats or -onlystats. You shouldn't need to use this option, really.
If the image file name is given as -, it means stdin
The image file can be raw row-major bytes (or 16-bits for
10- 12- and 16-bit images) with color planes stored one after
the other. In this case, the -height, -width, -planes, and -insubsamp
options must have been correctly set whenever they differ from
the default values. OR, the image can be in pgm or ppm format
in which csae none of these options need to be set as their values
will be read from the image file

** Image_flags can be:
	 -rgbtoycc: the input image is in RGB color format. these are
	              to be converted to YCrCb before use
	 -rgbto2ycc: convert RGB to YCbCr, subsample 2:1 in horiz and
	             vert directions for Cb and Cr planes


Command Interface
-----------------

After the optimization algorithm is done running, rdopt
prompts the user with

Command>

unless the -cmdfile option was used in which case commands are read
from the file named in that option. In response to most commands,
quantization tables Q (and thresholding tables, if -thresh was
specified) will be generated.

Valid commands are:

     [compress] size <target_size> 
       find Q to get compressed size <= target_size bytes

     [compress] bpp <target_bpp> 
       find Q to get compressed size <= target_bpp bits per pixel

     [compress] psnr <target_psnr> 
       find Q to get psnr >= target_psnr dB

     [compress] snr <target_snr> 
       find Q to get snr >= target_snr dB

     [compress] rmse <target_rmse> 
       find Q to get rmse <= target_rmse 

     qfile fname
       output subsequent next qtables generated by  commands to
       the file fname instead of the default file name (explained below)
     
     cfile fname
       output subsequent compressed files to the file fname instead
       of the default
     
     stats
       If this command is issued, all subsequent qtables will
       also contain coefficient-wise breakup of rate and distortion.
       This is useful for splitting the jpeg file into scans to
       use the "progressive" mode. This directory also contains
       a program (QTtoSF.c) which will convert this break-up info
       into a "scans file" that can be used by IJG's cjpeg. See below
       (in the section on supporting software) for details.
       Printing these stats can be disabled by specifying:
     
     nostats
       Do not print r-d breakups in qtable files (default).
     
     correct <use_bpp>
       You need cjpeg for this. The qtables found will be used to
       actually compress the image to get the actual bits per pixel
       (real_bpp). Subsequently, all reported bpp's will be corrected
       by the amount real_bpp - use_bpp. This can be disabled with:
     
     nocorrect
       Do not apply bpp corrections (default).

  A file RDOPT.Q.<command>.<target> will be generated in response
  to the commands "size", "bpp", "psnr", "snr", and "rmse"
  (unless a qfile command has been issued earlier), if the "compress"
  prefix is not given. This file will contain the quantization
  table(s) and the threshoding table(s) (if -thresh was used).
  If the "compress" prefix is specified, then  IJG's cjpeg will be
  used to compress the image according to the quantization table(s)
  found. The compressed file will have the name <imagename>.jpg,
  unless a cfile command has been issued earlier.

  In case the lagrangian method is used, three additional commands
  can be issued at the command interface:

    deltabpp <value>
      While searching for a target bpp, stop as soon as you
      have come within plusminus <value>. Default is 0.001, you
      can set this to zero.
    
    deltamse <value>
      While searching for a target distortion, stop as soon as you
      have come within plusminus <value> of the mse. 
      Default is 0.01, you can set this to zero.

    deltalambda <value>
      While using the bisection method to search for the 
      Lagrangian parameter lambda that will give you rate/distortion
      closest to the target, stop when the lambda interval
      becomes smaller than <value>. Default is 1e-8. Avoid tinkering
      with this. Especially, do not set this to zero! 


**********************************************************************
**********************************************************************
                   	    MORE INFO:
**********************************************************************
**********************************************************************

When should I use the Lagrangian method for optimization?
  Almost always. This would result in substantial improvement
  in speed. However, for thresholding tmax greater than about 
  100 (never really needed), use dynamic programming.

What is unit_num??
  Unit number is a number between 0 and (t-1) where t is the
  number of units in the image. Each unit is a set of color planes
  quantized with the same qtable. 

  Examples:

  rdopt -bpplane 0 -meth lagr -numtables 2 -im lena.ppm -rgbto2ycc

    Here, there are three color planes (Y, Cb, Cr) and
    two units. Unit # 0 consists of the Y plane and
    Unit # 1 consists of the Cb and Cr planes.

  rdopt -im lena.pgm

    Here, there is just one plane, and one unit.

  rdopt -meth lagr -planes 5 -numtables 3 -height 120 -width 120 -im some.raw.image

    Here, there are 5 color planes and 3 units. Unit # i has
    only the color plane # i, for i=0,1. Unit # 2 consists of
    color planes #2, #3, and #4.

Accuracy of results
-------------------

The psnr/snr/rmse values predicted by rdopt are *very* accurate.
The bpp predicted are *usually* within 0.02 bits per pixel of the
actual rate achieved by any good jpeg compressor using customized
huffman tables.

Using the generated Q tables
----------------------------

The Q tables can be used directly by the cjpeg program of the
Independent JPEG group. For example,

rdopt -im lena.pgm

Command> qfile tab1
Command> bpp 1.0
Command> quit

can be followed with

cjpeg -dct float -optimize -qtables tab1 -grayscale lena.pgm > lena.jpg

We recommend the use of -dct float and -optimize options in
the cjpeg program, for accuracy of the psnr/bpp predicted by
rdopt.

Using thresholding
------------------

For each unit, zeroing thresholds for DCT coefficients can be
determined too. For example,

rdopt -method lagrangian -thresh 20 -im lena.pgm

Command> qfile tab2
Command> bpp 1.0
Command> quit

This will produce the file tab2 which will look like:

#RDOPT.Qv2.0
#Image read from PGM (raw) file ../../../archive/lena2/IM
#Width x Height:512 x 512 / (1 x 1)
#Color-conversion applied: none
#Number of qtables used: 1
#Optimization parameters (method: Lagrangian): 
#        DC clamp = 255, ThreshSpan = 20
#
#Bits per pixel for each unit and the whole image
#    reported as num-bits/262144.000000, the denominator
#    being the sum of num pixels in all planes
#
#Quantization table for color planes 0 thru 0
9 10 9 9 9 10 11 11 
9 10 11 9 11 10 11 10 
9 10 11 11 10 11 12 10 
10 10 10 11 10 10 11 11 
11 10 11 10 10 12 11 10 
11 10 12 11 11 11 10 6 
11 10 10 12 10 11 11 12 
12 6 11 11 11 11 4 13 
# Table of thresholds
#T   4.5   5.5   5.5   5.5   6.0   6.5   7.5   8.0 
#T   4.5   6.0   6.5   6.0   7.0   6.5   7.5   7.5 
#T   6.5   6.0   7.0   7.0   7.5   7.5   8.0   8.0 
#T   6.5   6.5   7.0   7.5   7.0   7.5   8.0   8.5 
#T   7.5   7.0   7.0   7.0   7.5   8.0   8.0   8.5 
#T   7.5   7.5   8.0   7.5   8.5   8.5   8.5   9.5 
#T   8.5   8.5   8.0   9.0   8.5   9.0   9.0   9.5 
#T   9.0   9.5   9.5   9.0   9.0   9.0  10.0  11.5 
#Bpp = 1.000247 bits per pixel
#Size = 32776 bytes
#SNR = 33.584820 dB
#PSNR = 39.262947 dB
#RMSE = 2.775830
#END

The threshold table is given in lines beginning with
#T .......
To use the threshold table (and it'd better be used if you want
to achieve the specified rate/distortion!), you need to modify
your favorite JPEG compressor so that it can read these threshold
tables, and zero off every DCT coefficient that is less than the
corresponding threshold. IJG's cjpeg can be easily modified to do so:
see below and read the file cjpeg_thresh. Note that there will be
one threshold table for every unit.

*********************************************************************
*********************************************************************
                     SUPPORTING SOFTWARE
*********************************************************************
*********************************************************************

1. Independent JPEG Group's cjpeg
  
Use ver6 or higher, preferably. rdopt will try to fork off cjpeg
if "-correct" is used, or if a "compress" or "correct" command is
issued at the command interface. cjpeg does not support thresholding.
A simple modified version of cjpeg (to support thresholding) can be
built by following the instructions in the file "cjpeg_thresh". The
modified cjpeg can take the command line option "-thresh <tfile>"
tfile must contain thresholding table(s), on lines beginning
with "#T ". rdopt puts quantization and thresholding tables in the
same file, but lines containing  thresh tables begin with "#T ".

2. QTtoSF

This program converts coefficient-wise rate/distortion statistics
present in the qtable file rdopt produces (need to issue a "stats"
command at rdopt command interface for this) into progressive scan
breakups such that each AC scan will have approximately the same
size. The output is a 'scans file' that can be used by cjpeg
in its "-scans <file>" command line option.

Usage is: QTtoSF <numACscans> qtabfile [scanfile]

10- 12- and 16- bit images
--------------------------

rdopt can be compiled for these images too, by changing
SAMPLEBITS and QTABBITS in precision.h
These images can only be in the raw form, stored as
2 bytes per pixel.

For 12- and 16- bit images it's advisable to use -mapq
option for greater speed.

I haven't really tested rdopt much with 10, 12 or 16 bit
images. Comments/bug-reports etc. will be appreciated.


Comments, Suggestions, Bugs
----------------------------

Please send email to

ratnakar@cs.wisc.edu

Viresh Ratnakar
University of Wisconsin-Madison 
Computer Sciences Dept
1210 W Dayton St
Madison, WI 53706



CHANGES


RD-OPT ver2.0 has changed from ver1.0, in the following ways:
-------------------------------------------------------------

1. Bits per pixel (bpp) reported for images with more than
   one plane used to be
     total # of bits / (sum of # pixels in all planes)
   It still is the same, but now you can specify the option
     -bppplane n
   which would make the denominator the # of pixels in plane n.
   For example, for RGB images converted to YCrCb and subsampled
   in Cr and Cb, you can specify "-bppplane 0" to get the bpp
   as (#bits/(width x height)).
   ** "-bppplane 0" should probably be used always, as that is
      the conventional meaning of "bits-per-pixel"

2. Default DC clamp used to be 12, it is 255 now

3. Weighted PSNR: there was a bug which has been removed.
   Also, the coefficient  weights can be different in different
   units. So the option is now set as
     -weights n <weights_file>
   where n is the unit number

4. Unit weights: each unit can be weighed differently for
   calculating the average distortion using the option
     -pweights w0[,w1,..]

5. Optimization method: Used to be dynamic programming, which is
   still the default, but you can set it to lagrangian by
   specifying
     -method lagrangian
   Lagrangian will usually be faster, except for very high values
   in the -thresh option (see below). The "-bppscale" and
   "-bppmax" options will be ignored if the method is lagrangian.

6. Command interface:
     see the file BETTER_README.. essentially the same as before,
     but you can now compress from the command line if you have
     IJG's cjpeg. You can also set corrections to reported bpp
     using the actual bpp if you have cjpeg.

7. Thresholding:
     The main improvement: global zeroing-off levels for DCT
     coefficients can be determined, for better R-D performance.
     Use the option
       -thresh <tmax>
     This will generate thresholding tables for DCT coeffs, The
     higher the value of tmax, the greater the span of thresholds
     tried. All thresholds in the range [q/2 .. (q+tmax)/2] will
     be tried at every qtable entry q, in steps of 0.5
     See BETTER_README for details.

8. Bpp correction: there are two ways to specify this. You can
     use the "-correct <cbpp>" option, and/or issue any number
     of "correct <cbpp>" commands at the command interface. In
     either case, IJG's cjpeg will be used to compress the image 
     using the qtable(s) found for the rate <cbpp>, and subsequently
     reported bpp's will be correceted by the difference between
     the actual bpp and <cbpp>.

9. Plot: you can specify -pbppmax to set the max bpp that will
    be used for -plot. Default is 1.0.

10. -dcdpcm option: previously, by default an approximation of DPCM
   coding was used to estimate entropy of quantized and DPCM-coded
   DC coeffs. It is turned off be default now, but you can turn it
   on with this option. Without -dcdpcm, the distortion estimate is
   slightly more accurate, but it is not clear whether the rate 
   estimate is more or less accurate. In any case, the difference
   is usually so small that this can be swept under the rug.




src/rdopt.h

/*
** Copyright 1995,1996 by Viresh Ratnakar, Miron Livny
**
** Permission to use and copy this software and its documentation
** for any non-commercial purpose and without fee is hereby granted,
** provided that the above copyright notice appear in all copies and that
** both that copyright notice and this permission notice appear in
** supporting documentation.
**
**
** The University of Wisconsin and the copyright holders make no
** representations about the suitability of this software for any
** purpose.  It is provided "as is" without express or implied warranty.
**
**
** THE UNIVERSITY OF WISCONSIN AND THE COPYRIGHT HOLDERS DISCLAIM ALL
** WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES
** OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE UNIVERSITY OF
** WISCONSIN OR THE COPYRIGHT HOLDERS BE LIABLE FOR ANY SPECIAL, INDIRECT
** OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
** OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
** OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE
** OR PERFORMANCE OF THIS SOFTWARE.
**
** Author:  Viresh Ratnakar
** Email:   ratnakar@cs.wisc.edu
**
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "version.h"
#include "precision.h"

#define CQTAB "RDOPT.Q" VERSION
#define CQHIST "RDOPT.H" VERSION

#include "Image.h"

extern FILE *errfile;

/* quantization table entries in the range 1..QTABENTRYMAX */
#if (QTABBITS==8)
#define QTABENTRYMAX 255
typedef unsigned char Qentry;
#elif (QTABBITS==10)
#define QTABENTRYMAX 1023
typedef unsigned short Qentry;
#elif (QTABBITS==12)
#define QTABENTRYMAX 4095
typedef unsigned short Qentry;
#elif (QTABBITS==16)
#define QTABENTRYMAX 65535
typedef unsigned short Qentry;
#else
/** Force syntax error **/
QTABBITS must be 8,12,or16
#endif


#ifndef STRLENMAX
#define STRLENMAX 200
#endif

#ifndef boolean
#define boolean int
#endif
#ifndef TRUE
#define TRUE 1
#define FALSE 0
#endif

#define FFLOAT double

#define INFINITY ((FFLOAT) 1e300) /* machine-dependent */

#define ZERO_CORRECTION ((FFLOAT) ((MAXSAMPLE+1)<<2))

#define PLOT_POINTS 20 /* default # of points to plot with DumpPlot set */

#define NUM_HOOKS 20 /* default # of lamda intervals in
			lagrangian opt */
#include "opers.h"

/*********************************
type definitions
**********************************/

/******* histograms **********/
struct HistStruct
{
    /*
      PlusCount[v] will store the number of times the DCT
      coefficient (for which this struct is storing the
      counts) takes a postive value that gets Discretized
      to the integer v. (For definition of "Discretization"
      see opers.h)
      MinusCount[v] will store the number of times the DCT
      coefficient takes a negative value that gets Discretized
      to the integer -v.
      The array bounds will be 0..PlusSize and 0..MinusSize,
      respectively.
    */

    int PlusSize;
    int MinusSize;
    int *PlusCount;
    int *MinusCount;
};
typedef struct HistStruct Hist;



/************** OptimizationJob ***********/
struct OptimJobStruct
{
    Image TheImage;
    int NumTables; /* Number of quantization tables to be found.
		    If NumTables < TheImage.NumComponents, then
		    the last table will be used for all the
		    color components numbered (NumTables - 1)
		    thru (TheImage.NumComponents - 1). We call
		    each set of components being quantized together
		    a "Unit"
                 */

    boolean MapQ;
    boolean Silent;
    int VerboseLevel;
    boolean ReadCmdFromFile;
    char CmdFileName[STRLENMAX];

    int ThreshSpan;
    /* for qentry=q, thresholds upto
     q/2 + ThreshSpan/2 will be tried */

    boolean UseLagrangian;

    boolean useCorrection;
    FFLOAT correctionBpp;
    FFLOAT addToTableBpp;

    union
    {
        struct
        {
            /*** components to be used for dynamic programming  ***/

            FFLOAT BppMax;
            int BppScale;
            int TableRowSize;
            FFLOAT *LastRow[MAXCOMPONENTS]; /* last row of Dyn Prog table
  				     for each unit */
            int LeastLast[MAXCOMPONENTS]; /* lower and */
            int MostLast[MAXCOMPONENTS];  /* upper bounds on the extent of LastRow */
            FFLOAT *CombinedLastRow;
            int CombinedLeast;
            int CombinedMost;
            int *ToCombine[MAXCOMPONENTS]; /* ToCombine[i][bits] will give
  				    the starting point in LastRow[i] */

            Qentry *QChoice[MAXCOMPONENTS][64];
            Qentry *TChoice[MAXCOMPONENTS][64];

            FFLOAT *Err[MAXCOMPONENTS][64];
            int *Bpp[MAXCOMPONENTS][64]; /* BPP for each unit needs
  				     to be retained */
            /* explain the complexity of encoded use of Err and Bpp some day */
            boolean ErrEncodesRow[MAXCOMPONENTS][64];
            Qentry *QforBpp[64];
            Qentry *TforBpp[64];

            int BppOffsetInEncoding[MAXCOMPONENTS][64];
            int BppMaxInEncoding[MAXCOMPONENTS][64];


        }  dp;

        struct
        {
            /***** components for Lagrangian method  ***/
            FFLOAT ErrScale; /* each error will be divided by this */
            FFLOAT *Err[MAXCOMPONENTS][64];
            FFLOAT *Bpp[MAXCOMPONENTS][64];
            int    *SortedIndex[MAXCOMPONENTS][64];
            int     IndexEntries[MAXCOMPONENTS][64];
            FFLOAT *Lambda[MAXCOMPONENTS][64];
            FFLOAT LambdaMax;
            FFLOAT BppMax;
            FFLOAT BppMin;
            FFLOAT ErrMax;
            FFLOAT ErrMin;

            FFLOAT *BppHook;
            FFLOAT *LambdaHook;
            FFLOAT *ErrHook;
            int NumHooks;

        }  lagr;


    } opt_method;

    boolean WeightedComps;
    FFLOAT CompWeights[MAXCOMPONENTS];

    boolean WeightedCoefs[MAXCOMPONENTS];
    FFLOAT CoefWeights[MAXCOMPONENTS][64];

    FFLOAT bppWeight[MAXCOMPONENTS]; /* bpp = \sum_{i=0..NumTables-1}
					       bpp-for-ith-unit *
					       bppWeight[i] */
    FFLOAT RemainingSignal[MAXCOMPONENTS][64];
    FFLOAT LogSignalSq[MAXCOMPONENTS];
    FFLOAT LogTotalSignalSq;
    FFLOAT LogPeakSq[MAXCOMPONENTS];
    FFLOAT LogTotalPeakSq;
    FFLOAT NumPixels[MAXCOMPONENTS];
    FFLOAT NumBlocks[MAXCOMPONENTS];
    FFLOAT TotalNumPixels;
    FFLOAT NumPixelsForBpp;

    boolean ImFilePresent;
    boolean HistFilePresent;
    char HistFile[STRLENMAX];

    Hist Histogram[64];
    boolean DumpStats;
    boolean OnlyDumpStats;

    int MinTable[MAXCOMPONENTS][64];
    int MaxTable[MAXCOMPONENTS][64];
    int DCclamp;

    boolean UseDCDPCM;

    boolean DumpPlot;
    int PlotPoints;
    FFLOAT PlotBppMax;
    char PlotFileName[STRLENMAX];

    int bppplane; /* -1 implies all planes */


};

typedef struct OptimJobStruct OptimJob;


#define HEADER_BYTES 200
/** approximate number of header bytes needed for each unit **/



/********************** prototypes ************************/



extern void SetHistogram(Hist *H, Image *Im, int cnum, FFLOAT *mssq,
                         FFLOAT *SigTab, boolean UseDCDPCM);

extern void HistIncrCount(Hist *H, int coefnum, int val,
                          boolean waspositive);

extern void InitHistogram(Hist *H);
extern void ReadHistogram(FILE *hfile, Hist *H, FFLOAT *mssq);

extern void DumpHistogram(FILE *hfile, Hist *H, char *Description, FFLOAT mssq);

extern void FreeHistogram(Hist *H);

extern FFLOAT GetTotalBlocks(Hist *H);


extern void WriteDescription(char *s, Image *Im, int unum1, int unum2);

extern void InitOptimJob(OptimJob *job);

extern void GetParams(OptimJob *job, int argc, char *argv[]);

extern void Usage(void);

extern void BriefUsage(void);

extern boolean GetActualBpp(OptimJob *job, char *qfile, FFLOAT *ans);

extern boolean Compress(OptimJob *job, char *qfile, char *cfile, int *nbytes);

extern int newline(char s[], FILE *fp);
extern int newlinefd(char s[], int fd);

extern void FatalError(char *s);

extern void ReadIntTable(FILE *fp, int *tab);

extern void ReadRealTable(FILE *fp, FFLOAT *tab);

extern void WriteIntTable(FILE *fp, int *tab,  char *prefix);

extern void WriteThreshTable(FILE *fp, int *tab,  char *prefix);

extern boolean NotEqualReal(FFLOAT x, FFLOAT y);

extern void PrepareForErrBpp(OptimJob *job, int unum);

extern void lagrPrepareForErrBpp(OptimJob *job, int unum);

extern void SetErrBpp(OptimJob *job, int unum);

extern void lagrSetErrBpp(OptimJob *job, int unum);

extern void lagrSortErrBpp(OptimJob *job, int unum);

extern void Optimize(OptimJob *job, int unum);

extern void CombineUnits(OptimJob *job);

extern void Epilogue(OptimJob *job);

extern void lagrEpilogue(OptimJob *job);

extern void DumpJobChars(OptimJob *job, FILE *fp);

extern void Command(OptimJob *job);

extern void lagrCommand(OptimJob *job);

extern boolean lagrFindRate(OptimJob *job, FFLOAT *bpp, FFLOAT *err, FFLOAT delta, FFLOAT lambdamin, FFLOAT lambdamax, FFLOAT *lambda, int (*anstab)[64]);

extern boolean lagrFindErr(OptimJob *job, FFLOAT *bpp, FFLOAT *err, FFLOAT delta, FFLOAT lambdamin, FFLOAT lambdamax, FFLOAT *lambda, int (*anstab)[64]);

extern void WriteErrBppDist(FILE *fp, int bppscale, int *bpptab,
                            FFLOAT *errtab, char *prefix);

extern void WriteErrFBppDist(FILE *fp, FFLOAT *bpptab,
                             FFLOAT *errtab, char *prefix);


extern void SigToRemainingSig(FFLOAT *Sig, FFLOAT *RemainingSig);

extern void SetSignal(Hist *H, FFLOAT *Sig);

extern int ScanRowForErr(FFLOAT *row, FFLOAT err, int least, int most);

extern int ScanRowForBpp(FFLOAT *row, int bpp, int least, int most);

extern void GetStartingPoints(int *ToCombine[], int *StartingPt,
                              int Loc, int NumUnits);

extern void RecoverQ(int Loc, int unum, OptimJob *job, int *Q, int *Bpp,
                     FFLOAT *Err);

extern void RecoverQandT(int Loc, int unum, OptimJob *job, int *Q,
                         int *T, int *Bpp, FFLOAT *Err);

extern void lagrRecoverQandT(OptimJob *job, int (*QandT)[64],
                             int (*Q)[64], int (*T)[64], FFLOAT (*Bpp)[64], FFLOAT (*Err)[64]);

extern char * FileNameTail(char *s);


